### PR TITLE
[FEAT]: add numeric dot

### PR DESCRIPTION
**Pull request by universalmind303**

### CODE REVIEW

File: daft/expressions/expressions.py [hunk 1]
```diff
@@ -428,6 +428,16 @@ def ceil(self) -> Expression:
         expr = self._expr.ceil()
         return Expression._from_pyexpr(expr)
 
+    def dot(self, other: Expression) -> Expression:
```
Comment (samster25) [inline daft/expressions/expressions.py:431]: thoughts on also overloading `@` like in numpy?

```
In [6]: np.ones(10) @ np.ones(10)
Out[6]: 10.0
```

File: daft/daft [hunk 2]
```diff
@@ -0,0 +1 @@
+../Daft/daft
```
Comment (samster25) [inline daft/daft:1]: what does this file do?

File: src/daft-dsl/src/functions/numeric/dot.rs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    fn evaluate(&self, inputs: &[Series], _: &FunctionExpr) -> DaftResult<Series> {
+        if inputs.len() != 2 {
+            return Err(DaftError::ValueError(format!(
+                "Expected 2 input args, got {}",
+                inputs.len()
+            )));
+        }
+        let input = inputs.first().unwrap();
+        let other = inputs.last().unwrap();
```
Comment (samster25) [inline src/daft-dsl/src/functions/numeric/dot.rs:41]: let's add a TODO for making this a faster dot product kernel

File: src/daft-dsl/src/functions/numeric/dot.rs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    fn to_field(&self, inputs: &[ExprRef], schema: &Schema, _: &FunctionExpr) -> DaftResult<Field> {
+        if inputs.len() != 2 {
+            return Err(DaftError::SchemaMismatch(format!(
+                "Expected 2 input arg, got {}",
+                inputs.len()
+            )));
+        }
+        let field = inputs.first().unwrap().to_field(schema)?;
```
Comment (samster25) [inline src/daft-dsl/src/functions/numeric/dot.rs:23]: how would this handle the case if the types we're not the same.
```
int32 dot f32
```

File: daft/daft (hunk 2 above)
Comment (universalmind303) [reply daft/daft:1]: oh whoops i didn't mean to add this.

File: daft/expressions/expressions.py (hunk 1 above)
Comment (universalmind303) [reply daft/expressions/expressions.py:431]: what about nested/list dtypes? do we just throw an unsupported error if you try to matmul on list types? 

numpy appears to error
```py
a = [[1, 2], [3, 4]]
b = [[5, 6], [7, 8]]

print(a @ b)
# TypeError: unsupported operand type(s) for @: 'list' and 'list'
```

Comment (universalmind303) [reply daft/expressions/expressions.py:431]: would we want to support nested dtypes too? 

```py
a = np.array([[1, 2], [3, 4]])
b = np.array([[5, 6], [7, 8]])

print(a @ b)
# [[19 22]
# [43 50]]

**Comment by universalmind303** [comment]
hmm.. If i merge/rebase my branch, it changes the default behavior. 

```py
before = df.select(col("a").dot(col("b"))collect().to_pydict()['a']
# [32]

after
# [32,32,32]
``` 

I would expect a `.select(expr)` with a single aggregate expr to be evaluated as an aggregate, but now it's broadcasting it back to the original len 🤔. I'm trying to cherrypick what commit changed this behavior.

**Comment by universalmind303** [comment]
so i think this is actually blocked by https://github.com/Eventual-Inc/Daft/issues/1979

**Pull request closed without merging**